pkg/chains/formats/slsa/v2alpha4/internal/taskrun/taskrun.go

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright 2024 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package taskrun
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	slsa "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v1"
+	"github.com/tektoncd/chains/pkg/artifacts"
+	"github.com/tektoncd/chains/pkg/chains/formats/slsa/extract"
+	buildtypes "github.com/tektoncd/chains/pkg/chains/formats/slsa/internal/build_types"
+	externalparameters "github.com/tektoncd/chains/pkg/chains/formats/slsa/internal/external_parameters"
+	internalparameters "github.com/tektoncd/chains/pkg/chains/formats/slsa/internal/internal_parameters"
+	resolveddependencies "github.com/tektoncd/chains/pkg/chains/formats/slsa/internal/resolved_dependencies"
+	"github.com/tektoncd/chains/pkg/chains/formats/slsa/internal/slsaconfig"
+	"github.com/tektoncd/chains/pkg/chains/objects"
+)
+
+const (
+	taskRunResults     = "taskRunResults/%s"
+	taskRunStepResults = "stepResults/%s"
+)
+
+// GenerateAttestation returns the provenance for the given taskrun in SALSA 1.0 format.
+func GenerateAttestation(ctx context.Context, tro *objects.TaskRunObjectV1, slsaConfig *slsaconfig.SlsaConfig) (interface{}, error) {
+	bp, err := byproducts(tro)
+	if err != nil {
+		return nil, err
+	}
+
+	bd, err := getBuildDefinition(ctx, slsaConfig.BuildType, tro)
+	if err != nil {
+		return nil, err
+	}
+
+	att := intoto.ProvenanceStatementSLSA1{
+		StatementHeader: intoto.StatementHeader{
+			Type:          intoto.StatementInTotoV01,
+			PredicateType: slsa.PredicateSLSAProvenance,
+			Subject:       extract.SubjectsFromBuildArtifact(ctx, tro),
+		},
+		Predicate: slsa.ProvenancePredicate{
+			BuildDefinition: bd,
+			RunDetails: slsa.ProvenanceRunDetails{
+				Builder: slsa.Builder{
+					ID: slsaConfig.BuilderID,
+				},
+				BuildMetadata: metadata(tro),
```

Comment:
Instead of defining a new metadata func below, what if we just use the one from `v1alpha3`? That way, there is no duplication of the functions.

Same comment for wherever we have the exact same information. I think this probably includes things like `externalParameters`, `internalParameters` etc.